fix: WDIO Cucumber Test Tags

sauravdas1997 opened the pull request:
## Proposed changes

[//]: # (Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue.)

Currently Browserstack accessibility is not running when tags are passed in accessibilityOptions in case of cucumber, so we are handling the use case for test tags in cucumber.

## Types of changes

[//]: # (What types of changes does your code introduce to WebdriverIO?)
[//]: # (_Put an `x` in the boxes that apply_)

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

## Checklist

[//]: # (_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._)

- [x] I have read the [CONTRIBUTING](https://github.com/webdriverio/webdriverio/blob/main/CONTRIBUTING.md) doc
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] I have added proper type definitions for new commands (if appropriate)

## Further comments

[//]: # (If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...)

### Reviewers: @webdriverio/project-committers

erwinheitzman commented:
Hi @sauravdas1997 thank you for your contribution!

I have some requests:
1. file an issue that can be coupled to this PR so that we can first verify and discuss if this is an issue
2. describe the changes made
3. add some tests to verify the changed behaviour

Thank you

erwinheitzman commented:
Hi @sauravdas1997 thank you for your contribution!

This does indeed need to be `name`, however the value also shouldn't be an array, could you make this change and update/add tests were needed? Thank you.

nextlevelbeard commented on packages/wdio-browserstack-service/tests/util.test.ts:
```diff
@@ -1026,13 +1026,30 @@ describe('validateCapsWithA11y', () => {
     })
 
-    it('returns false if full test name contains excludeTags', async () => {
+    it.only('returns false if full test name contains excludeTags', async () => {
```
Don't forget to remove this `.only`

sauravdas1997 replied:
done

sauravdas1997 commented:
> file an issue that can be coupled to this PR so that we can first verify and discuss if this is an issue
This is a fix for a browserstack specific offering that we came across during our internal testing.
> describe the changes made
Added in the PR description
> add some tests to verify the changed behaviour
Added in PR

erwinheitzman commented on packages/wdio-browserstack-service/src/util.ts:
```diff
@@ -343,11 +343,20 @@ export const validateCapsWithA11y = (deviceName?: any, platformMeta?: { [key: st
+            const tagsList: string[] = []
+            world?.pickle?.tags.map((tag: { [key: string]: any; }) => tagsList.push(tag.name))
+            const excluded = excludeTags?.some((exclude) => tagsList.includes(exclude))
+            const included = includeTags?.length === 0 || includeTags?.some((include) => tagsList.includes(include))
```
Could you explain why we do an explicit check here `includeTags?.length === 0`? I feel like the include and exclude arrays should be treated the same way so it's a surprise to see this.

sauravdas1997 replied:
Use case: If the user has not specified any includeTags, we should return the shouldScanTestForAccessibility results as true. However, if the user has mentioned a value in includeTags, we should evaluate against it. If the user does not provide excludeTags, no tags should be excluded by default. We will only honour it if the user explicitly specifies tags in excludeTags.

erwinheitzman reviewed (commented):
Got one question as a comment, looks got overall!

christian-bromann reviewed (approved):
LGTM 👍 

@erwinheitzman any objections to merge?

christian-bromann commented:
I will go ahead and merge. Happy to address remaining issues in a new PR.
